### ISSUE TITLE
Unsetting "hardware_reservation_id" does not trigger device re-creation

### ISSUE THREAD

**Issue by invidian** [packethost/terraform-provider-packet#289]
### Terraform Version
```
$ terraform -v
Terraform v0.13.5
+ provider registry.terraform.io/packethost/packet v3.1.0
```

### Affected Resource(s)
- `packet_device`

### Terraform Configuration Files
```diff
terraform {
  required_providers {
    packet = {
      source  = "packethost/packet"
      version = "~> 3.1.0"
    }
  }
}

resource "packet_device" "main" {
  hostname = "mat-test-reservation-id"
-  hardware_reservation_id          = "bar"
+ //hardware_reservation_id          = "bar"
  project_id                       = "foo"
  facilities                       = ["ams1"]
  operating_system                 = "ubuntu_16_04"
  plan                             = "c2.medium.x86"
  billing_cycle                    = "hourly"
  wait_for_reservation_deprovision = true
}

```

### Expected Behavior

When hardware reservation ID is no longer set, device should be re-created with on-demand instance.

### Actual Behavior

Terraform plan is empty and device is not re-created.

### Steps to Reproduce
1. `terraform apply`
2. Apply config change as noted in the diff above.
3. `terraform plan`

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- kinvolk/lokomotive#1156

**Comment by invidian** [comment]
I ended up with panicking SDK while testing this: https://github.com/hashicorp/terraform-plugin-sdk/issues/642

I _think_ this issue comes from the fact that combination of `Optional`, `Computed` and `ForceNew` is used, which cannot really be well handled from Terraform SDK point of view. One solution would be to split `hardware_reservation_id` field into one which is user-configured and other which is a computed attribute, which can be read e.g. when `next-available` is used.

**Comment by t0mk** [comment]
@invidian thanks for reporting, I will take a look tomorrow. I know the hw res id attr is problematic. It also needs to support the `next-available` value, which makes it a bit hard to interpret, use, and process in the code.

**Comment by t0mk** [comment]
@invidian I did try your use case, and it's true, if you create a device resource with hw res, and then remove the hw res attr, the device is not marked for re-creation.

I noticed the hardware_reservation_id is marked as "Computed", and when I unset the "Computed" flag, it attempts re-creation after removing the hw res attr from HCL template. However, I think the Computed is somehow necessary for the "next-available", functionality, because I set it when adding support for "next-available" in this commit: https://github.com/packethost/terraform-provider-packet/commit/e3da606156e110147d784e363fb550c4c4c7d1d0

I will test if the next-available would somehow work without the Computed flag.

**Comment by t0mk** [comment]
Maybe I could get around the next-available, if I'd introduced another attr called "deployed_hardware_reservation_id", which would be Computed. Similarly to "deployed_facility" https://github.com/packethost/terraform-provider-packet/blob/master/packet/resource_packet_device.go#L62

Maybe we could then even get rid off the DiffSuppressFunc.

**Comment by t0mk** [comment]
Hmm, just removing "Computed: true" from the hardware_reservation_id was enough to seemingly make it work as expected. I am not sure if it's an OK move, and whether it will break some other use cases. Nothing comes to my mind though. I will read a bit about Computed.

**Comment by invidian** [comment]
As far as I know, without `Computed: true` you cannot use `Set()` in `CustomizeDiff`, but this is not used here, so it shouldn't be a problem.

As for `hardware_reservation_id` `DiffSuppressFunc` is used, I think it should be OK to remove it, as Terraform won't try to revert `Set()` from `Read` function.

I think test case below represents the situation. Tests are passing so I _think_ it should be OK.
```go
package main

import (
        "context"
        "testing"

        "github.com/hashicorp/terraform-plugin-sdk/v2/diag"
        "github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
        "github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
)

func readF(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
        if err := d.Set("hardware_reservation_id", "something"); err != nil {
                panic(err)
        }

        return nil
}

func TestImportTest_import(t *testing.T) {
        resource.Test(t, resource.TestCase{
                IsUnitTest: true,
                Providers: map[string]*schema.Provider{
                        "import": {
                                ResourcesMap: map[string]*schema.Resource{
                                        "import_test": {
                                                CreateContext: func(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
                                                        d.SetId("foo")

                                                        return readF(ctx, d, m)
                                                },
                                                ReadContext: readF,
                                                UpdateContext: func(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
                                                        return nil
                                                },
                                                DeleteContext: func(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
                                                        d.SetId("")

                                                        return nil
                                                },
                                                Schema: map[string]*schema.Schema{
                                                        "dummy_field": {
                                                                Type:     schema.TypeString,
                                                                Required: true,
                                                        },
                                                        "hardware_reservation_id": {
                                                                Type:     schema.TypeString,
                                                                Optional: true,
                                                                ForceNew: true,
                                                                DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
                                                                        if new == "next-available" && len(old) > 0 {
                                                                                return true
                                                                        }
                                                                        return false
                                                                },
                                                        },
                                                },
                                        },
                                },
                        },
                },
                Steps: []resource.TestStep{
                        {
                                Config: `
resource "import_test" "foo" {
  dummy_field             = "dhg"
  hardware_reservation_id = "next-available"
}
`,
                                Check: resource.TestCheckResourceAttr("import_test.foo", "hardware_reservation_id", "something"),
                        },
                        {
                                Config: `
resource "import_test" "foo" {
  dummy_field             = "dhg"
  hardware_reservation_id = "next-available"
}
`,
                                PlanOnly:           true,
                                ExpectNonEmptyPlan: false,
                                Check:              resource.TestCheckResourceAttr("import_test.foo", "hardware_reservation_id", "something"),
                        },
                },
        })
}
```

**Comment by t0mk** [comment]
@invidian thanks for the test snippet, that's really inventive way to verify TF features. It's first time I see it and I'm gonna use it more.

The next-available will most likely not be a problem, and the Computed shouldn't have been in the hw_res_id attribute. Later on, I became a bit worried about resources that might link the `hardware_reservation_id` attr from the device (to refer to the hw res for any reason), so I added a Computed field `deployed_hardware_reservation_id` in #293.

**Comment by displague** [comment]
This has been resolved in [v3.2.0](https://github.com/packethost/terraform-provider-packet/releases/tag/v3.2.0)
